clang/tools/driver/CacheLauncherMode.cpp

```diff
@@ -30,15 +33,54 @@ static bool isSameProgram(StringRef clangCachePath, StringRef compilerPath) {
   return CacheStat.getUniqueID() == CompilerStat.getUniqueID();
 }
 
+static bool shouldCacheInvocation(ArrayRef<const char *> Args,
+                                  IntrusiveRefCntPtr<DiagnosticsEngine> Diags) {
+  SmallVector<const char *, 128> CheckArgs(Args.begin(), Args.end());
+  // Make sure "-###" is not present otherwise we won't get an object back.
+  CheckArgs.erase(
+      llvm::remove_if(CheckArgs, [](StringRef Arg) { return Arg == "-###"; }),
+      CheckArgs.end());
+  CreateInvocationOptions Opts;
+  Opts.Diags = Diags;
+  std::shared_ptr<CompilerInvocation> CInvok =
+      createInvocation(CheckArgs, std::move(Opts));
+  if (!CInvok)
+    return false;
+  if (CInvok->getLangOpts()->Modules) {
+    Diags->Report(diag::warn_clang_cache_disabled_caching)
+        << "-fmodules is enabled";
+    return false;
+  }
+  if (CInvok->getLangOpts()->AsmPreprocessor) {
+    Diags->Report(diag::warn_clang_cache_disabled_caching)
+        << "assembler language mode is enabled";
+    return false;
+  }
+  return true;
+}
+
 Optional<int>
 clang::handleClangCacheInvocation(SmallVectorImpl<const char *> &Args,
                                   llvm::StringSaver &Saver) {
   assert(Args.size() >= 1);
 
+  IntrusiveRefCntPtr<DiagnosticOptions> DiagOpts;
+  if (Optional<std::string> WarnOptsValue =
+          llvm::sys::Process::GetEnv("LLVM_CACHE_WARNINGS")) {
+    SmallVector<const char *, 8> WarnOpts;
+    WarnOpts.push_back(Args.front());
+    llvm::cl::TokenizeGNUCommandLine(*WarnOptsValue, Saver, WarnOpts);
+    DiagOpts = CreateAndPopulateDiagOpts(WarnOpts);
+  } else {
+    DiagOpts = new DiagnosticOptions();
+  }
   auto DiagsConsumer = std::make_unique<TextDiagnosticPrinter>(
-      llvm::errs(), new DiagnosticOptions(), false);
-  DiagnosticsEngine Diags(new DiagnosticIDs(), new DiagnosticOptions());
+      llvm::errs(), DiagOpts.get(), false);
+  IntrusiveRefCntPtr<DiagnosticsEngine> DiagsPtr(
+      new DiagnosticsEngine(new DiagnosticIDs(), DiagOpts));
+  DiagnosticsEngine &Diags = *DiagsPtr;
   Diags.setClient(DiagsConsumer.get(), /*ShouldOwnClient=*/false);
+  ProcessWarningOptions(Diags, *DiagOpts, /*ReportDiags=*/false);
```

Comment:
Why not report issues with the diagnostic options?

Reply:
This was copy-pasting from another place, seems like a good idea to report, updated 👍